Pod events: add reason for volume mount timeout

tsmetana opened the pull request:
If a fsGroup for a pod is being set, the volume needs to be traversed and all the files get chown-ed. This may take a long time and cause a mount timeout events to be emitted for the pod. The patch adds some more information to the events so the user knows why the volume is still not mounted.
```release-note
NONE
```

tsmetana committed:
Pod events: add reason for volume mount timeout

If a fsGroup for a pod is being set, the volume needs to be traversed
and all the files get chown-ed. This may take a long time and cause a
mount timeout events to be emitted for the pod. The patch adds some more
information to the events so the user knows why the volume is still not
mounted.

childsb commented:
/approve

davidz627 commented on pkg/kubelet/volumemanager/cache/actual_state_of_world.go:
```diff
@@ -323,6 +330,41 @@ func (asw *actualStateOfWorld) RemoveVolumeFromReportAsAttached(volumeName v1.Un
+	asw.Lock()
+	defer asw.Unlock()
+
+	volumeObj, volumeExists := asw.attachedVolumes[volumeName]
```
is it ok to just ignore if volume doesn't exist?

tsmetana replied:
It's not possible to SetUp a volume that is not marked as attached in the ASW (it shouldn't be even mounted).

davidz627 replied:
so why check volumeExists? If it doesn't exist shouldn't we throw an error? I would prefer that than having this function fail silently

tsmetana replied:
OK then. I'll fix this.

davidz627 commented on pkg/kubelet/volumemanager/cache/actual_state_of_world.go:
```diff
@@ -323,6 +330,41 @@ func (asw *actualStateOfWorld) RemoveVolumeFromReportAsAttached(volumeName v1.Un
+	defer asw.Unlock()
+
+	volumeObj, volumeExists := asw.attachedVolumes[volumeName]
+	if volumeExists {
```
is it ok to just ignore if volume doesn't exist?

tsmetana replied:
dtto. If the volume is not attached, it should be unmounted so the fact it is SetUp doesn't matter.

davidz627 commented on pkg/kubelet/volumemanager/volume_manager.go:
```diff
@@ -380,6 +382,27 @@ func (vm *volumeManager) WaitForAttachAndMount(pod *v1.Pod) error {
 
+// Figure out the volume's unique name to find out whether it's still being set-up
+func (vm *volumeManager) isVolumeBeingSetUp(outerVolumeSpecName string, volumesToMount []cache.VolumeToMount) bool {
+	for _, volumeToMount := range volumesToMount {
```
instead of looping through all "volumesToMount" for each unmountedVolume, you could loop less and just create a set/list of volumes being set up, right now I think there's redundant looping

tsmetana replied:
The looping happens only in the case the volume mount times-out. This is not a standard situation, I'm not sure keeping another list in the memory to speed up spitting an error message is worth it.

tsmetana replied:
OK. I don't have strong preference here so I will implement it as you suggest.

davidz627 commented on pkg/volume/util/operationexecutor/operation_generator.go:
```diff
@@ -540,8 +540,10 @@ func (og *operationGenerator) GenerateMountVolumeFunc(
+		actualStateOfWorld.MarkVolumeAsBeingSetUp(volumeToMount.VolumeName)
 		// Execute mount
 		mountErr := volumeMounter.SetUp(fsGroup)
+		actualStateOfWorld.MarkVolumeAsSetUp(volumeToMount.VolumeName)
```
I dont know if "VolumeSetUp" is the right naming here. Because here you call it whether we error or not, so in the error case the volume isnt "SetUp" per-se.

Maybe "VolumeInSetUp" or "VolumeSetUpPending" is more apt?

tsmetana replied:
MarkVolumeAsSetUp was not meant to bear the information about the result of the operation, just the fact it finished.  So what about "MarkVolumeAsSetUpFinished"? "SetUpPending" sounds like we're waiting for it to start which is not true.

davidz627 commented on pkg/kubelet/volumemanager/volume_manager.go:
```diff
@@ -368,6 +368,8 @@ func (vm *volumeManager) WaitForAttachAndMount(pod *v1.Pod) error {
 			return nil
 		}
 
+		vm.addUnmountedVolumesStatus(unmountedVolumes)
```
silly question: where do these status' get consumed/exposed?

tsmetana replied:
The status being printed in the pod events ('kubectl describe pod'). The whole point of the patch is to calm the user down when the pod event log contains mount time-out messages. In the case the pod is being set up, ie. the recursive chown takes long time there's no way to speed it up and the mount needs to be re-tried again (the volume is actually mounted on the system but must not be mounted to the pod yet) so the time-out does not mean anything wrong happened in such case.

The pod events are easily accessible so it looked like a good place to expose the information.

davidz627 replied:
Thanks for the explanation, sounds good to me

davidz627 commented on pkg/kubelet/volumemanager/cache/actual_state_of_world.go:
```diff
@@ -205,6 +208,7 @@ type actualStateOfWorld struct {
 	// volumePluginMgr is the volume plugin manager used to create volume
 	// plugin objects.
 	volumePluginMgr *volume.VolumePluginMgr
+
```
nit: extra space

tsmetana replied:
Ah, sorry. Will remove it.

davidz627 reviewed (changes requested):
Just some questions and small comments

tsmetana commented:
@davidz627 Thanks for the review.

tsmetana commented:
Sorry for the delays... I have added error logs when trying to mark a volume that does not exist, renamed the method but did not do anything about the looping. I was trying to figure out how to match the outerVolumeSpecName with UniqueVolumeName and could not find anything that would actually not hinder the code readability even more.

Actually I noticed there is an unrelated bug in the kubelet's volume manager in the getUnattachedVolumes where the volume names are also confused resulting in reporting all the expected volumes as unattached. I will create a separate PR for that.
